lc0/src/neural/network_opencl.cc

```diff
@@ -0,0 +1,212 @@
+/*
+    This file is part of Leela Chess Zero.
+    Copyright (C) 2017 Gian-Carlo Pascutto
+    Copyright (C) 2018 The LCZero Authors
+
+    Leela Zero is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Leela Zero is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Leela Zero.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "network_blas_cl.h" // factory.h, network.h, optionsdict.h
+#include "utils/random.h"
+#include <cstdio>
+
+namespace lczero {
+
+OpenCLNetwork::OpenCLNetwork(const Weights& weights, const OptionsDict& options)
+  : BlasNetwork(weights, options) {
+  // ^ nontrivial: all cpu initialization is shared by OpenCL initialization
+  // this function corresponds to Network.cpp:418-496
+  printf("Initializing OpenCL.\n");
+  auto channels = weights_.input.biases.size();
+  opencl_.initialize(channels);
+
+  for (auto& opencl_net : opencl_.get_networks()) {
+    auto tuners = opencl_net->getOpenCL().get_sgemm_tuners();
+    auto mwg = tuners[0];
```

Comment:
Do you know what `mwg`, `kwg`, `vwm` are? :) If yes, would be nice to add comment.

Reply:
not a goddamn clue! and lack of comments are one of my biggest frustration with leela zero.

Reply:
Part of the Winograd convolution is matrix matrix multiplication SGEMM. For the best performance on GPU it calculates multiplication in tiles. The optimal size of the tiles is GPU specific and the best parameters for the current GPU are searched by the tuner. Again for the best performance we want the matrix sizes to be divisible by the required tile size. MWG, KWG and VWM are parameters that decide the tile size and are needed here to zero pad the weight matrices to the correct size.